pkg/broker/catalog.go

```diff
@@ -2,149 +2,100 @@ package broker
-				size{Name: "M40"},
-				size{Name: "M50"},
-				size{Name: "M60"},
-				size{Name: "M80"},
-				size{Name: "M200"},
-				size{Name: "M300"},
-			},
-		},
-		cloud{
-			Name: "AZURE",
-			Sizes: []size{
-				size{Name: "M10"},
-				size{Name: "M20"},
-				size{Name: "M30"},
-				size{Name: "M40"},
-				size{Name: "M50"},
-				size{Name: "M60"},
-				size{Name: "M80"},
-				size{Name: "M200"},
-			},
-		},
-	}
-}
+// findProviderAndInstanceSizeByIDs will search all available providers and
+// instance sizes to find the ones matching the specified service and plan ID.
+func (b Broker) findProviderAndInstanceSizeByIDs(serviceID, planID string) (*atlas.Provider, *atlas.InstanceSize, error) {
+	for _, providerName := range providerNames {
+		provider, err := b.atlas.GetProvider(providerName)
+		if err != nil {
+			return nil, nil, err
+		}
 
-func cloudFromPlan(serviceID string, planID string) (*cloud, *size) {
-	for _, cloud := range clouds() {
-		if cloud.ID() == serviceID {
-			for _, size := range cloud.Sizes {
-				if size.ID(cloud) == planID {
-					return &cloud, &size
+		if serviceIDForProvider(provider) == serviceID {
+			for _, instanceSize := range provider.InstanceSizes {
+				if planIDForInstanceSize(provider, instanceSize) == planID {
+					return provider, &instanceSize, nil
 				}
 			}
 		}
 	}
 
-	return nil, nil
+	return nil, nil, errors.New("invalid service ID or plan ID")
 }
 
-// cloud represents a single available cloud in which a cluster can be deployed.
-type cloud struct {
-	Name  string
-	Sizes []size
-}
-
-// Plans generates service broker plans based on the available instance sizes.
-func (c cloud) Plans() []brokerapi.ServicePlan {
-	plans := make([]brokerapi.ServicePlan, len(c.Sizes))
+// plansForProvider will convert the available instance sizes for a provider
+// to service plans for the broker.
+func plansForProvider(provider *atlas.Provider) []brokerapi.ServicePlan {
+	var plans []brokerapi.ServicePlan
 
-	for i, size := range c.Sizes {
-		plans[i] = brokerapi.ServicePlan{
-			ID:          size.ID(c),
-			Name:        size.Name,
-			Description: fmt.Sprintf("Instance size \"%s\"", size.Name),
+	for _, instanceSize := range provider.InstanceSizes {
+		plan := brokerapi.ServicePlan{
+			ID:          planIDForInstanceSize(provider, instanceSize),
+			Name:        instanceSize.Name,
+			Description: fmt.Sprintf("Instance size \"%s\"", instanceSize.Name),
 		}
+
+		plans = append(plans, plan)
 	}
 
 	return plans
 }
 
-// ID generates a globally unique service ID for use in the catalog.
-func (c cloud) ID() string {
-	return fmt.Sprintf("%s-service-%s", idPrefix, strings.ToLower(c.Name))
-}
-
-// size represents a single instance size which clusters can use.
-// TODO: Add memory and storage to generate better plan descriptions.
-type size struct {
-	Name string
+// serviceIDForProvider will generate a globally unique ID for a provider.
+func serviceIDForProvider(provider *atlas.Provider) string {
```

Comment:
Would it maybe make sense to make this a function on the Provider object? e.g.
```go
func (provider *atlas.Provider) getServiceID string {
```

Reply:
I'm trying to keep the atlas package totally separate from the broker. As the service ID is a service broker concept I felt it made more sense to keep it in the broker package. Does that make sense?